core/src/main/java/org/apache/calcite/sql/type/OperandTypes.java

```diff
@@ -496,6 +496,71 @@ public static SqlOperandTypeChecker variadic(
   public static final SqlSingleOperandTypeChecker LITERAL =
       new LiteralOperandTypeChecker(false);
 
+  /**
+   * Operand type-checking strategy type must be a boolean non-NULL literal.
+   */
+  public static final SqlSingleOperandTypeChecker BOOLEAN_LITERAL =
+      new FamilyOperandTypeChecker(ImmutableList.of(SqlTypeFamily.BOOLEAN),
+          i -> false) {
+        @Override public boolean checkSingleOperandType(
+            SqlCallBinding callBinding,
+            SqlNode operand,
+            int iFormalOperand,
+            boolean throwOnFailure) {
+          if (!LITERAL.checkSingleOperandType(
+              callBinding,
+              operand,
+              iFormalOperand,
+              throwOnFailure)) {
+            return false;
+          }
+
+          if (!super.checkSingleOperandType(
+              callBinding,
+              operand,
+              iFormalOperand,
+              throwOnFailure)) {
+            return false;
+          }
+
+          final SqlLiteral arg = (SqlLiteral) operand;
+          final boolean isBooleanLiteral =
+              SqlLiteral.valueMatchesType(arg.getValue(), SqlTypeName.BOOLEAN);
+
+          if (!isBooleanLiteral) {
+            if (throwOnFailure) {
+              throw callBinding.newError(
+                  RESOURCE.argumentMustBeBooleanLiteral(
+                      callBinding.getOperator().getName()));
+            }
+            return false;
+          }
+          return true;
+        }
+      };
+
```

Comment:
comments:
refers to NUMERIC_UNIT_INTERVAL_NUMERIC_LITERAL

the reason why i am not define the type like bellowing also is this

```
public static final SqlFunction SORT_ARRAY =
    SqlBasicFunction.create(SqlKind.SORT_ARRAY,
        ReturnTypes.ARG0_NULLABLE,
        OperandTypes.ARRAY.or(
            OperandTypes.sequence("'SORT_ARRAY(<ARRAY>, <BOOLEAN>)'",
                OperandTypes.ARRAY, OperandTypes.BOOLEAN_LITERAL)));
```

Reply:
There's already a `BOOLEAN` entry in `OperandTypes`, what about this function necessitates the need for `BOOLEAN_LITERAL`?

Reply:
@tanclary it is need, you can look the unit test i add , which will fail in spark

f.checkFails("^sort_array(array[2, null, 1], cast(1 as boolean))^",
        "Argument to function 'SORT_ARRAY' must be a literal", false);
and sort_array(array[2, null, 1], b) , b is column